### PR TITLE
Adds support to read and write to Databricks delta tables

### LINKED ISSUE #611: [FEA]: Support reading data from delta lake tables
### Is this a new feature, an improvement, or a change to existing functionality?

New Feature

### How would you describe the priority of this feature request

Medium

### Please provide a clear description of problem this feature solves

I would like morpheus to read from delta lake tables in one of the stages. As delta lakes are widely used as source of data from streaming sources and batch sources, it would be great to support reading data directly from these tables for the morpheus pipelines reducing the time taken to load the data through some other mechanism. 

### Describe your ideal solution

A new stage that would take in the parameters for databricks/delta table either through function args or OS Env variables

### Describe any alternatives you have considered

Have tried the work around of reading the data from S3 in other format, but the time taken to load the data can be too high making the total runtime of the pipeline way too high

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I have searched the [open feature requests](https://github.com/nv-morpheus/Morpheus/issues?q=is%3Aopen+is%3Aissue+label%3A%22feature+request%22%2Cimprovement%2Cenhancement) and have found no duplicates for this feature request

### PR DESCRIPTION
**Pull request by pthalasta**
Adds support for querying data from Databricks delta tables and write the results to delta tables. Resolves #611

### CODE REVIEW

File: docker/Dockerfile
Comment (mdemoret-nv) [reply docker/Dockerfile:48]: Looks like there are openjdk conda packages. Should investigate if we can use these.

File: morpheus/stages/output/write_to_deltalake_stage.py
Comment (mdemoret-nv) [inline morpheus/stages/output/write_to_deltalake_stage.py:209]: This function is repeated now in multiple places. It should be pulled out of the stages into a common location.

File: morpheus/stages/input/deltalake_source_stage.py
Comment (mdemoret-nv) [inline morpheus/stages/input/deltalake_source_stage.py:75]:
```suggestion
        return "from-deltalake"
```
Match the command name

**Review by mdemoret-nv — CHANGES_REQUESTED** [review]
Suggestions are inline. Biggest issue is that there are no tests that exercise these new stages. To guarantee that they will continue to work, we need tests built that run these stages.

File: docker/Dockerfile
Comment (pdmack) [reply docker/Dockerfile:48]: @mdemoret-nv strongly agree
probably need a different and possibly layered strategy here

**Comment by pthalasta** [comment]
will take a look at the changes

**Comment by BartleyR** [comment]
@pthalasta Pinging here to resurface. This would be a good stage to add once issues are addressed.

File: docker/Dockerfile
Comment (BartleyR) [reply docker/Dockerfile:48]: @mdemoret-nv @pthalasta Are we closer on a resolution or path here? What's the update?

**Comment by pthalasta** [comment]
@mdemoret-nv i can resolve the comments except the usage of JDK as it is a requirement and regardless of which source we install it from there will be increase in container size. As this stage reads data from pyspark via databricks-connect, i don't think it will work without JDK. There are early stage python libs to read from delta without any involvement of databricks/spark, if JDK alternative is a must, i can probably test with early phase python libs, please let me know.

**Comment by gurpreets-nvidia** [comment]
@cwharris @pthalasta @mdemoret-nv : what are next steps with this MR? Are any discussions happening offline to find a way to resolve the JDK issue and close/merge this MR?

**Comment by cwharris** [comment]
@gurpreets-nvidia we are indeed discussions options. It looks like adding the JRE, JDK, and Databricks Connect will introduce an additional ~300MB to the release container. We're currently in the process of determining what CVE's would be introduced. Once we understand the CVE implications we can determine whether including it is the right way to go or if we need to look for a workaround.

**Comment by pthalasta** [comment]
@cwharris i see two options
- Load delta table into memory using python based delta table libs and then filter data the data. This doesn’t require spark so no, JDK. However, there can be memory issues in certain cases if we try to load all the data without filtering
- Pre-filter data using spark or distributed compute that can read the delta tables and share the resulting df as done in the approach above and use the result df as needed in downstream jobs.

Let me know how to proceed in closing on this MR

File: morpheus/stages/input/deltalake_source_stage.py
Comment (mdemoret-nv) [reply morpheus/stages/input/deltalake_source_stage.py:106]: Is this path going to be relative to the user's HOME directory? If so, this can be replaced with `config_file = "~/.databricks-connect"`

Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py:106]: yes, those changes are being made in the next PR

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py:86]: yes, this is intended for single time execution and doesn't include pagination

File: morpheus/stages/input/deltalake_source_stage.py
Comment (mdemoret-nv) [reply morpheus/stages/input/deltalake_source_stage.py:106]: Next PR? You can just push updates to this branch and it will update this PR

Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py:106]: yes, thats what i mean, sorry for the confusion.

**Comment by cwharris** [comment]
@pthalasta Looks like there are no new CVEs introduced with the JRE/JDK + databricks-connect. :tada: 

I'm still getting confirmation that we are going to include this in the release container, but it sounds like if/when we do, it will be a newer (probably 11) LTS version of the JRE/JDK. This will present a warning with the databricks-connect API, as it indicates it is not supported with Java>8, but I can't imagine why that would be the case, given that version of Java is from 2014.

**Comment by cwharris** [comment]
In the meantime, would it be possible to address the review comments @mdemoret-nv has made?

**Comment by pthalasta** [comment]
yes, i'm working on the changes, should have it next week

File: docker/Dockerfile
Comment (cwharris) [inline docker/Dockerfile:48]: Do we need both the JRE and JDK, or will just the JRE work? Also, let's update this to use the version `11` to match with what we're using for development, and use `headeless` to keep the footprint small.
```suggestion
    DEBIAN_FRONTEND=noninteractive apt-get install -y openjdk-11-jre-headless &&\
```

Comment (pthalasta) [reply docker/Dockerfile:48]: sure, will make those changes, i'm testing it with only jre atm, will update when i've things working

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: @mdemoret-nv paginating has been added in new commit, resolving this, please review new commits

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: @mdemoret-nv changes have been made in new commit

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: @mdemoret-nv  please review new commit

File: docker/Dockerfile
Comment (pthalasta) [reply docker/Dockerfile]: @cwharris made changes in new commit, please review

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: @mdemoret-nv sorry, i did not understand, do you mean there is no need for additional message?

File: docker/conda/environments/cuda11.8_runtime.yml
Comment (dagardner-nv) [inline docker/conda/environments/cuda11.8_runtime.yml:22]: Not sure if this is needed

File: docker/conda/environments/cuda11.8_examples.yml
Comment (dagardner-nv) [inline docker/conda/environments/cuda11.8_examples.yml:37]: This dep doesn't need to be in both the dev and examples yaml files. 
Just the dev yaml should be fine

Comment (pthalasta) [reply docker/conda/environments/cuda11.8_examples.yml:37]: sounds good, i'll make the changes

File: docker/conda/environments/cuda11.8_dev.yml
Comment (dagardner-nv) [inline docker/conda/environments/cuda11.8_dev.yml:105]: If your installing this as a pip package, it needs to be underneath the `- pip:` line. 
If it is available as a conda package, then it should sorted alphabetially.

Comment (pthalasta) [reply docker/conda/environments/cuda11.8_dev.yml:105]: yes, the changes are in my local will be pushed by EOD for the same

File: morpheus/utils/databricks_utils.py
Comment (cwharris) [inline morpheus/utils/databricks_utils.py:29]: There's a couple of things going on here that we'll want to change.

First, we're prioritizing the host environment variable over the local function parameter. We'll want to switch that so the application can override the host. This:
```python
"host": os.environ.get('DATABRICKS_HOST', databricks_host),
```
becomes this:
```python
"host": os.environ.get('DATABRICKS_HOST') if databricks_host is None else databricks_host,
```

Second, the `None` checks in the first part of the function look a little complicated and also very similar to the line above. Let's try to simplify it and combine the logic at the same time. This:
```python
    if (os.environ.get('DATABRICKS_HOST', None) is None and databricks_host is None):
        raise Exception("Parameter for databricks host not provided")
```
can become:
```python
    if databricks_host is None:
        databricks_host = os.environ.get('DATABRICKS_HOST')
        if databricks_host is None:
            raise Exception("Parameter for databricks_host not provided")
```
Then we're able to remove the logic from the `config = { ... }` and just use the values directly:
```python
    config = {
        "host": databricks_host,
        ...
    }
```

File: morpheus/utils/databricks_utils.py
Comment (cwharris) [inline morpheus/utils/databricks_utils.py:44]: We won't be calling this function very often, so eliminating unnecessary writes isn't something we need to worry about. With that considered, we can simplify this code by removing the file read and comparison.
```suggestion
    with open(config_file, "w") as fh:
        fh.write(json.dumps(config))
    logger.info("Databricks-connect successfully configured!")
```

**Review by cwharris — CHANGES_REQUESTED** [review]
Thanks for the updates. Pradeep. There's a few more changes we need to make to get this PR ship shape before merging. Looking forward to getting this PR in after those.

File: morpheus/stages/output/write_to_databricks_deltalake_stage.py
Comment (pthalasta) [reply morpheus/stages/output/write_to_databricks_deltalake_stage.py:94]: let me check as these are global spark configs rather than that of session configs. I'll look into spark docs and get back

File: morpheus/utils/databricks_utils.py
Comment (cwharris) [reply morpheus/utils/databricks_utils.py:44]: Can you be more specific? From what I'm reading, it appears that the logic is "if the file exists and contains the contents we would write, do not write the file. Otherwise, write the file." I don't see how there are two different sets of credentials in the config. Did you mean to append to the file with `a` instead of overwrite it with `w+`? If so, I think this logic needs some more work, as simply appending could lead to more problems.

File: morpheus/utils/databricks_utils.py
Comment (dagardner-nv) [reply morpheus/utils/databricks_utils.py:29]: Thinking about this a little more, these are the same environment variables used to configure `morpheus/modules/mlflow_model_writer.py`, and it appears to be the only way to configure the model writer.

I'm assuming these may need to be configured independently.

File: morpheus/utils/databricks_utils.py
Comment (pthalasta) [reply morpheus/utils/databricks_utils.py]: @cwharris the condition and other changes that were missed to be checked in should come in during next push, but the process is to check if the configuration exists, then we skip writing or making changes to the file else, we append the configs to the file

File: morpheus/utils/databricks_utils.py
Comment (pthalasta) [reply morpheus/utils/databricks_utils.py]: @dagardner-nv are you suggesting that there is an existing config/module that needs to be reused?

File: docker/conda/environments/cuda11.8_dev.yml
Comment (pthalasta) [reply docker/conda/environments/cuda11.8_dev.yml]: changes have been made

File: morpheus/stages/output/write_to_databricks_deltalake_stage.py
Comment (pthalasta) [reply morpheus/stages/output/write_to_databricks_deltalake_stage.py]: changes have been made

File: morpheus/stages/output/write_to_databricks_deltalake_stage.py
Comment (pthalasta) [reply morpheus/stages/output/write_to_databricks_deltalake_stage.py]: changes have been made

File: morpheus/stages/output/write_to_databricks_deltalake_stage.py
Comment (pthalasta) [reply morpheus/stages/output/write_to_databricks_deltalake_stage.py]: changes have been made

File: morpheus/stages/output/write_to_databricks_deltalake_stage.py
Comment (pthalasta) [reply morpheus/stages/output/write_to_databricks_deltalake_stage.py]: changes have been made

File: docker/conda/environments/cuda11.8_runtime.yml
Comment (pthalasta) [reply docker/conda/environments/cuda11.8_runtime.yml]: changes have been made

File: morpheus/utils/databricks_utils.py
Comment (pthalasta) [reply morpheus/utils/databricks_utils.py]: Changes with `a` has been made, i'm testing the overall function to check for issues, will push in changes once its been tested

File: morpheus/utils/databricks_utils.py
Comment (cwharris) [reply morpheus/utils/databricks_utils.py]: @pthalasta I believe David is suggesting the use of unique environment variable names for this stage specifically because our mlflow implementation already uses (some of) these specific environment variable names. With this in mind, I don't think it makes sense to have the environment variables override the application-specific settings, because the application can declare more than one stage. We will need to rethink how we approach this problem in the general case. For the time being, @pthalasta please update your code according to my [original comment](https://github.com/nv-morpheus/Morpheus/pull/630#discussion_r1315148842), and we can address the issue at a later point.

Comment (cwharris) [reply morpheus/utils/databricks_utils.py]: @dagardner-nv Issue created here: https://github.com/nv-morpheus/Morpheus/issues/1174

File: morpheus/utils/databricks_utils.py
Comment (cwharris) [reply morpheus/utils/databricks_utils.py]: What is the format of the configuration? I imagine it would be json, but the way you are appending configs makes it seem as though the format is json-lines. If it is not json-lines, this will present a deserialization problem if more than one config is added to the file.

File: morpheus/utils/databricks_utils.py
Comment (cwharris) [reply morpheus/utils/databricks_utils.py]: @pthalasta Can you also make sure that the environment variables are prepended with `DELTA_TABLE_`? This will disambiguate them from the mlflow variables.

Comment (cwharris) [reply morpheus/utils/databricks_utils.py]: After talking it over with the team, we _don't_ want the `DELTA_TABLE_` prefix. Sorry. :grimacing:

**Comment by pthalasta** [comment]
@cwharris @dagardner-nv  while working on the latest changes to update databricks-connect to v13 that helps passing credentials directly to databricks session object (than having to configure as global env variables) i'm encountering the following error while converting Spark df to cudf (with pandas as intermediate step i.e sparkDF->pandasDF->cudf)

```
Exception occurred in pipeline. Rethrowing
Traceback (most recent call last):
  File "/home/jovyan/utils/delta_spark_source_stage.py", line 110, in source_generator
    yield MessageMeta(df=cudf.from_pandas(df.toPandas().drop(["_id"],axis=1)))
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/nvtx/nvtx.py", line 101, in inner
    result = func(*args, **kwargs)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/cudf/core/dataframe.py", line 7491, in from_pandas
    return DataFrame.from_pandas(obj, nan_as_null=nan_as_null)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/nvtx/nvtx.py", line 101, in inner
    result = func(*args, **kwargs)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/cudf/core/dataframe.py", line 5119, in from_pandas
    data[col_name] = column.as_column(
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/cudf/core/column/column.py", line 2188, in as_column
    data = as_column(pa.Array.from_pandas(arbitrary), dtype=arb_dtype)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/cudf/core/column/column.py", line 1976, in as_column
    col = ColumnBase.from_arrow(arbitrary)
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/cudf/core/column/column.py", line 374, in from_arrow
    result = libcudf.interop.from_arrow(data)[0]
  File "interop.pyx", line 178, in cudf._lib.interop.from_arrow
RuntimeError: CUDA error encountered at: /opt/conda/conda-bld/work/cpp/src/bitmask/null_mask.cu:93: 200 cudaErrorInvalidKernelImage device kernel image is invalid

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/envs/morpheus/lib/python3.10/site-packages/morpheus/pipeline/pipeline.py", line 327, in join
    await self._mrc_executor.join_async()
  File "/home/jovyan/utils/delta_spark_source_stage.py", line 113, in source_generator
    raise Exception(e)
Exception: CUDA error encountered at: /opt/conda/conda-bld/work/cpp/src/bitmask/null_mask.cu:93: 200 cudaErrorInvalidKernelImage device kernel image is invalid
```

any idea how i can fix this or what might be causing this?

**Comment by dagardner-nv** [comment]
> any idea how i can fix this or what might be causing this?

Some things to try, first ensure all of the LFS objects are populated by running:
```
./scripts/fetch_data.py fetch all
```

If that doesn't fix the issue, ensure that your conda environment is up to date with the latest changes, specifically in this case with cudf. 

Failing that, post a repro command.

**Comment by cwharris** [comment]
@pthalasta

That error seems to indicate that the kernels have been build for a different architecture of GPU than the one you are running. 
- What GPU are you using on the machine that produced the error?
- How did you obtain/compile Morpheus and cuDF?

**Comment by pthalasta** [comment]
@cwharris i'm running the container on kubeflow with Tesla GPUs. In order to use it with kubeflow i used morpheus runtime image and start command of jupyter lab for the container

**Comment by cwharris** [comment]
@pthalasta cuDF (and therefore Morpheus) does not support Kepler GPUs. Is that your problem?

**Comment by pthalasta** [comment]
we are using Tesla V100 and also A10

**Comment by cwharris** [comment]
@pthalasta Are you rebuilding Morpheus or changing the conda environment from inside the container?

**Comment by pthalasta** [comment]
@cwharris just adding additional packages to the conda env, no change to morpheus build not rebuilding morpheus as the image is a runtime images used as base

**Comment by cwharris** [comment]
@pthalasta 
- What does `nvidia-smi` report?
- What does `conda list | grep cudf` report?

**Comment by pthalasta** [comment]
@cwharris my bad, its Tesla T4s

output of nvidia-smi

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.161.03   Driver Version: 470.161.03   CUDA Version: 11.8     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            On   | 00000000:00:1E.0 Off |                    0 |
| N/A   45C    P8    16W /  70W |      0MiB / 15109MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

for conda output

```
# conda list | grep cudf
cudf                      23.06.01        cuda11_py310_230629_g6a548b072d_0    rapidsai
cudf_kafka                23.06.01        py310_230629_g6a548b072d_0    rapidsai
dask-cudf                 23.06.01        cuda11_py310_230629_g6a548b072d_0    rapidsai
libcudf                   23.06.01        cuda11_230629_g6a548b072d_0    rapidsai
libcudf_kafka             23.06.01        230629_g6a548b072d_0    rapidsai
```

**Comment by cwharris** [comment]
I was wrong. This is not the error associated with a mismatched device/build architecture. I am not sure what the error is caused by, but it sounds like the kernel may be corrupt somehow. Are you able to reproduce this error in a fresh container?

**Comment by pthalasta** [comment]
@cwharris yes, i see the same error in new container as well

**Comment by pthalasta** [comment]
@cwharris any update on how i can resolve the issue? I'm blocked on moving ahead with implementation and testing because of the error

**Comment by cwharris** [comment]
Sorry @pthalasta , I'm unfamiliar with the environment you're launching in and I haven't experiences this error first hand. I'll work with the other members on the team to see if we can repro the error. Any additional information you can provide about your environment would be helpful in determining the cause of these errors.

**Comment by pthalasta** [comment]
@cwharris can you take a look at the latest changes? If it looks good, i'll move ahead with working on test cases.

File: morpheus/utils/databricks_utils.py
Comment (pthalasta) [reply morpheus/utils/databricks_utils.py]: @cwharris this config is removed with latest changes, resolving the convo

File: morpheus/utils/databricks_utils.py
Comment (pthalasta) [reply morpheus/utils/databricks_utils.py]: marking as resolved as there are no changes needed here

File: morpheus/stages/output/write_to_databricks_deltalake_stage.py
Comment (pthalasta) [reply morpheus/stages/output/write_to_databricks_deltalake_stage.py]: have made these changes in the latest code

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: env variables have been removed with latest code. resolving

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: this code has been removed as per latest version changes of databricks-connect. Resolving

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: config util code has been removed as per latest changes. Resolving

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: have made the changes in the latest code

**Comment by cwharris** [comment]
@pthalasta changes look good to me. There are some formatting issues that CI will probably complain about, so I'll have it run and we can see what needs to change.

**Comment by cwharris** [comment]
@pthalasta changes look good to me. There are some formatting issues that CI will probably complain about, so I'll have it run and we can see what needs to change.

File: morpheus/stages/input/deltalake_source_stage.py
Comment (pthalasta) [reply morpheus/stages/input/deltalake_source_stage.py]: Marking as resolved as this has been changed to logger.error

**Comment by pthalasta** [comment]
@cwharris i think we have all tests and checks passing. There was no change to documentation part, so not sure if that is needed?

**Comment by pthalasta** [comment]
@dagardner-nv i see that the documentation is failing due to import of stages by sphinx. Is this expected?

**Comment by pthalasta** [comment]
@mdemoret-nv @cwharris can we merge if there are no other changes required and all looks good?

**Review by cwharris — APPROVED** [review]

**Review by mdemoret-nv — APPROVED** [review]

**Pull request merged**
